### ISSUE TITLE
Large World - Performance Optimizations

### ISSUE THREAD

**Issue by andrew-sidereal** [tainicom/Aether.Physics2D#61]
Hi Nikos,

I'm using R.U.B.E. to create world and load them into Aether.Physics2D. It's going well until I reach a couple hundred bodies and then it falls down to around 13fps, even with drawing disabled. 

It's interesting, as in R.U.B.E.'s simulator it runs very nicely at 60fps. I wonder how R.U.B.E. is so much faster. 

My aim is to create a HUGE physics world, with many thousands of bodies. I have multiplayer synchronization working with smaller worlds. I'm thinking I'll unload any bodies which aren't near to a player and represent them just with a single Vector2 placeholder. Every second or so I'll see if there are any placeholders within a player's interest area and load them up, or conversely, unload them if they've left. 

That said, since even these couple hundred bodies are running slowly, more optimizations will be required. I plan on using dotTrace to profile and attempt some optimizations. However, before delving in myself, I wanted to reach out to you and see if you had any ideas or guidance. Also, would you be interested in collaborating with me in this effort of creating a HUGE game world? (I also plan on converting from floats to double or decimal to allow for solar-system-scale worlds. (https://stackoverflow.com/questions/50847019/farseer-box2d-increasing-max-world-size-by-using-doubles-or-decimals-rather)

Long post, apologies. Feel free to contact me directly on email, if it's easier: alundgren04 (at) gmail.

**Comment by BlizzCrafter** [comment]
> It's going well until I reach a couple hundred bodies and then it falls down to around 13fps, even with drawing disabled.

From my experience this is a rather uncommon behavior. 

I created a test world on the Windows platform and had no problem reaching 60fps with several 1000th **active** dynamic bodies (Multithreading On).

On which platform are you trying to create your world?

You should also try to create your world without importing it via R.U.B.E. and see if it makes a difference.

**Comment by damian-666** [comment]
From the data you report, looks like IsBullet is on, for all the bodies, you might check that next.   A few hundred items all marked as bullet will stress the physics engine, piling of bullet invoking CCD to prevent object tunnelling will be slow with this engine and might not be necessary for your simulation

**Comment by nkast** [comment]
Hi @alundgren04 ,

I am not surprised as Box2d is C++ and the C++ compilers are very good at optimizing code. I tested the testbed app under .net 4.6.1 which uses the new RyuJIT clr but didn't notice any significant improvement. 

I am waiting to hear your findings with the profiler if you found any hotspots. So far I am mostly using the TumblerTest for profiling which focuses mostly on multiple contacts but with one fixture per body. What is your world like? Maybe you hit some other type of bottleneck.

Here some things you can try, 
Build with .net 4.6.1, 64bit, Release. I found that running the app through VS, even on release build, behaves differently than running the app stand alone.

Enable the multithreading solvers.
```
            World.ContactManager.VelocityConstraintsMultithreadThreshold = 256;
            World.ContactManager.PositionConstraintsMultithreadThreshold = 256;
            World.ContactManager.CollideMultithreadThreshold = 256;
```
Note that some callback events are fired by different threads. You can pick and choose which one you want enabled.

Adjust AABBExtension and LinearSlop. This depends on the size of your objects. The goal is to avoid frequent rebalance in the broadphase.

Reduce the values for VelocityIterations, PositionIterations, TOIVelocityIterations, TOIPositionIterations, MaxTOIContacts in Settings. (too low values might affect the stability of your engine)


About using double, unless there's a easy way to provide builds for both float and double i'd rather keep using floats as this is mainly a game engine. One option is to use structs in a way similar than C++ typedefs to wrap a float,double or fixed point int. But I don't know how this affects performance and what to avoid. I was thinking for example to strong type several values like rotation, mass, inertia, etc is a  way similar to TimeSpan/DateTime.
 a.g. Complex.Angle could returns an Angle struct, Complex.Magnitute a Length struct.
and provide custom operations like `static Momentum operator *(Mass value1, Velocity2 value2)`, 
`static Position2 operator +(Position2 value1, Velocity2 value2)`, ` static Distance2 operator -(Position2 value1, Position2 value2)`.

**Comment by andrew-sidereal** [comment]
Thank you, all for your time. I found the root issue with my test scenario, which was the fact that bodies were being imported and duplicated several times into the exact same position and with the same fixtures. Although the total number of bodies wasn't very high, it seems when bodies overlap exactly it leads to some serious performance issues, but this was easily resolved once identified. I'm going to continue building out test cases and seeing how far I can push the engine. I'll continue sharing my results here. I typically create the test case in RUBE and then import it into the Testbed.WINDOWS.MG project as a new test. 

I can provide you the RUBE loader code and RUBE files/JSON so you can be viewing the exact same test scenarios, if you like.

**Comment by andrew-sidereal** [comment]
I've created a couple test simulations and have two which seem to demonstrate a performance issue. I create a world 10,000m by 10,000m and fill it with a simple 1m x 1m single-fixture box with random velocity and spin. These are spaced one per 100m and results in over 13k bodies. Performance is impressive at just over 3ms.

Then I run the exact same simulation, but use bodies with 14 1m x 1m fixtures. Performance grids down to 43ms per update. 

Granted, this is a lot more fixtures, but given that there's a fair amount of space between them and they're not colliding often, and not at all for the first 10sec or so, I feel there is room for improvement here. I wasn't able to spin up a profiler tonight, but wanted to share the test cases with you all. 

These pictures demonstrate it well:
https://imgur.com/a/kRTGlAS

In both tests, multithreading is enabled and it's in release mode. Settings are all defaults.

Is there spatial partitioning implemented in Aether? I feel assigning bodies to grid cells and only checking those in nearby cells may help, but this may already be happening.

**Comment by nkast** [comment]
That's interesting. How did that perform under RUBE/Box2d?

> Is there spatial partitioning implemented in Aether? I feel assigning bodies to grid cells and only checking those in nearby cells may help, but this may already be happening.

Box2D has a broadphase that checks AABBs for potential collisions. Interestingly, Farseer added an interface layer that let's you swap broadphase with yous own implementation and at some point had an experimental quadtree broadphase.  
The broadphase  sends those data to the narrowphase where new collisions are formed if necessary.
The broadphase does not know about bodies, each fixture is linked to an AABB and the narrowphase is the one that filter those things out.


I suspect that here the broad-phase is overwhelmed. The lost of precision at 10.000m might also be at play. Certain things might trigger a rebalance of the broad-phase tree which can be very slow. You can't really know until you run a profiler.

Some things you can try:
 Adjuct the  AABBExtension, LinearSlop & AngularSlop.
 If possible, combine fixtures into larger ones to have less fixtures per body. 
 Make a smaller world around the center (say 500x500) that runs well (<10ms), then offset that near the 10000 boundary (say + [5000;5000]) to see if precision affects this in any way.

**Comment by nkast** [comment]
Do you have that test on gihhub on somewhere where we can see it?

**Comment by andrew-sidereal** [comment]
Holyfuzz, seems to have identified a similar shortcoming with multi-fixture bodies. His idea of creating one AABB for the entire body during BroadPhasing, rather than one for each fixture, makes a lot of sense to me. http://community.monogame.net/t/aether-physics2d-release-v-1-0/9709/20

@nkast I'm interested to hear your thoughts on this. I feel it should first just test the entire body, then dive into the fixtures, if needed. Is this flawed reasoning, or something worth delving into?

**Comment by waltdestler** [comment]
@alundgren04 I think it depends on your use case. The downside is that then each body needs its own DynamicTree of fixtures, which adds a bit of memory overhead per body, plus the overhead of having to frequently convert AABBs between coordinate systems. In my game, the performance improvements of putting whole body AABBs in the broad phase far outweighed that extra overhead, but your results may differ.

In an ideal world, I think the engine would dynamically determine per-body whether to put the individual fixtures in the broad-phase or to put the whole body in the broad phase, based on some experimentally-determined threshold number of fixtures attached to the body.

If anyone wants to see my modified Farseer code, [send me an email](https://waltdestler.com/contact.html).

**Comment by nkast** [comment]
from my reply in that discussion.

```
Have you tried to add a reference to the body in FixtureProxy? 
You could then quickly discard the pair inside .AddPair() without having to get the bodies through fixture, which probably cause a cache-miss.
```

Let me expand on that:
Add a `Body Body;` field in FixtureProxy.
Set `proxy.body = this.Body;` in `Fixture.CreateProxies(...)`.
in AddPair change the code to this,

```
            FixtureProxy proxyA = BroadPhase.GetProxy(proxyIdA);
            FixtureProxy proxyB = BroadPhase.GetProxy(proxyIdB);

            Body bodyA = proxyA.Body;
            Body bodyB = proxyB.Body;

            // Are the fixtures on the same body?
            if (bodyA == bodyB)
            {
                return;
            }
            
            Fixture fixtureA = proxyA.Fixture;
            Fixture fixtureB = proxyB.Fixture;

            int indexA = proxyA.ChildIndex;
            int indexB = proxyB.ChildIndex;
                        
            ...
```
With this change we no longer need to fetch Fixtures from ram into the cache.

If this works and you see some gains, then you can move it up to the broadphase.
Save a body reference next to the AABB (as an Object), and test with `ReferenceEquals()` to skip the AABB test between proxies. This should directly attach the issue and it's less work than a hierarchical test.

**Comment by andrew-sidereal** [comment]
I've created a fork of this repo here: https://github.com/alundgren04/Aether.Physics2D/commits/master

I'm working on migrating my performance tests to the existing TestBed project. I have one there currently, which can be improved, but works. You can alter the constants at the top of the file to alter the world size and body density:

`        const float WORLD_SIDE_SIZE = 10000f;`
`         const float METERS_PER_BODY = 100f;`

You may also change the body from a basic 1-fixture box to a complex body with many fixtures by changing this constant:

`var selectedBodyIndex = BOX_BODY_INDEX;`

I plan on making the IBroadPhase solver injectable via the world constructor and allow recreating the world with different solvers, different sizes, body densities, and body structures (box, complex, mix), in order to easily do multiple benchmark tests back-to-back without having to recompile. The reason for making the IBroadPhase injectable is so I can test out the performance effect of different implementations, such as the changes discussed in this thread.

That said, I'm flying out to Iceland later today and may not be able to get back to this until next week. I just wanted to get it out there ASAP.

Cheers!

**Comment by damian-666** [comment]
thanks for sharing this.  The old ODE engine allowed various spaces like HashSpace and QuadTree space with xforms to be collided together and i've seen nested body spaces trees give more than one order of magnitude performance increase in an earlier project with asteroid-like rigid bodies.

  Ideally, a more general solution could allow a broadphase culling of first just the outer AABB of a ---system of bodies connected with joints, not just fixtures on a body.   So systems of bodies like articulated creatures could be processed faster with just the update of the AABBs of the whole thing, and then using the AABB overlap area when creatures are fighting close.

**Comment by nkast** [comment]
I managed to have a quick look at it, 
In the first case (1 fixture per body) most time is spend in Fixture.Synchronize() and Body.SynchronizeFixtures(). I will see if I can get some local optimizations there.
When I used the second body with multiple fixtures, additional time was spend on BroadPhase.UpdatePairs(), almost as much as the time spend on Body.SynchronizeFixtures().

In a previous message you said, that the world run at 3ms with 13k bodies x 1 fixture/body and at 43ms with 13k bodies x 14 fixture/body. In terms of Fixtures (or Shapes to be exact), that's 4334 fixtures/ms vs 4233 fixture/ms. That's a fair comparison IMO.

**Comment by andrew-sidereal** [comment]
Hi @nkast any word on those Synchronize optimizations you identified?

**Comment by nkast** [comment]
Not yet but It's on my list to investigate this.

**Comment by nkast** [comment]
@alundgren04 
I posted some local optimizations in PR https://github.com/tainicom/Aether.Physics2D/pull/62. 
Can you confirm what performance you see?

**Comment by andrew-sidereal** [comment]
Hi @nkast, sorry for the delayed reply, but I wanted to get some solid benchmarks before replying. It's looking good. Please see the before and after here: https://imgur.com/a/qWN9aGx

Next, I'm going to attempt to do the broadphase calculation off of body AABB and only doing fixture AABB testing if the body AABBs overlap. I feel this will be a tremendous optimization for complex bodies, anecdotally and logically.

**Comment by nkast** [comment]
In my benchmark I saw a ~10% improvement using the SparseBodiesTest. 
Your tests have a much better improvement that I wouldn't expect from this type of local-optimization.
Anyways,  we don't disagree that this is an improvements, I still need to test Circle shapes and see I can apply similar improvements to the PR and then it's good to merge.

**Comment by nkast** [comment]
@alundgren04 
I just pushed PR #65. This will improve Data locality of TreeNode as node structs are packed closer together. It also consumes less memory and it's friendlier to the GC as there are no longer hundred or thousand of objects to track.

In the SparseBodiesWithManyFixturesTest I found that a significant portion of CPU cycles was spend in the inner loop of DynamicTree.Query(). Mainly on AABB.TestOverlap().
https://github.com/tainicom/Aether.Physics2D/blob/69ad2ed4addd9a9ac8713ee1441f8da9f94d9094/Physics2D/Collision/DynamicTree.cs#L349-L364

One step further would be to split the struct into Hot/Cold data.
```
AABB[] _nodes_AABB;                           // nodes Enlarged AABB.
TreeNodeChildred[] _nodes_Childred; // nodes { Child1, Child2 }.
TreeNode[] _nodes; // nodes { Height, Parent }.
T[] _nodes_UserData; // nodes UserData.
```
